### ISSUE TITLE
Create a separate function for evaluation, decouple train step.

### ISSUE THREAD

**Issue by cshewmake2** [sophiaas/torch-tools#2]
Currently, `Experiment::train_step` is called for three different purposes:
- training
- validation
- test evaluation

These three functions really should be distinct in general. Merging them into one function creates complexity and possibility for confusion with respect to true training, logging, or models in which the validation procedure is different from the training procedure.

I propose that Experiment and inherited classes of Experiment contain two virtual functions: **train_step** and **evaluate**:

**train_step** can be used for gradient updates with training data
**evaluate** can be used for validation or test set results

This becomes relevant especially if we're interested in logging **model parameters** during training, in which case we don't want to log a result for train, val, and test (as the model parameters don't change across sets, only data does).
